src/adapter/src/catalog.rs

```diff
@@ -1983,6 +1973,27 @@ impl Source {
             | DataSourceDesc::Source => None,
         }
     }
+
+    /// The number of persist shards that this source is _directly_ responsible for creating.
+    ///
+    /// For example, a PostgresSQL source only consumes a single shard, as do each  of its
+    /// subsources.
+    pub fn persist_shard_count(&self) -> i64 {
```

Comment:
```suggestion
    /// The expensive resource that each source consumes is persist shards. To
    /// prevent abuse, we want to prevent users from creating sources that use an
    /// unbounded number of persist shards. But we also don't want to count
    /// persist shards that are mandated by teh system (e.g., the progress
    /// shard) so that future versions of Materialize can introduce additional
    /// per-source shards (e.g., a per-source status shard) without impacting
    /// the limit calculation.
    pub fn user_controllable_persist_shard_count(&self) -> i64 {
```